Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli

```diff
@@ -188,41 +188,90 @@ struct ClipmapLevel
     float m_weight; // Blending weight between 2 clipmap layers.
 };
 
-// Calculate the most detailed clipmap level.
-ClipmapLevel CalculateClosestClipmapLevel(
-    float2 distanceFromViewPosition,
-    float maxRenderDistance,
-    float clipmapScaleBase,
-    uint clipmapStackSize
-)
+// returns 1.0 for true, 0.0 for false.
+float AabbContainsPoint(float2 center, float extent, float2 point_)
+{
+    float2 aabbMin = center - float2(extent, extent);
+    float2 aabbMax = center + float2(extent, extent);
+
+    // Use (x - aabb.x1)(x - aabb.x2) < 0 for collision detection.
+    float2 minDiff = point_ - aabbMin;
+    float2 maxDiff = point_ - aabbMax;
+    float2 det = float2(minDiff.x * maxDiff.x, minDiff.y * maxDiff.y);
+    float2 contains = step(float2(0.0, 0.0), det);
+    // x axis | y axis | result we want | x - 1 | y - 1 | (x - 1)(y - 1)
+    //    0   |   0    |      1         |   -1  |   -1  |        1
+    //    0   |   1    |      0         |   -1  |    0  |        0
+    //    1   |   0    |      0         |    0  |   -1  |        0
+    //    1   |   1    |      0         |    0  |    0  |        0
+    contains = contains - float2(1.0, 1.0);
+    return contains.x * contains.y;
+}
+
+ClipmapLevel CalculateMacroClipmapLevel(float2 worldPosition)
 {
+    float clipmapSize = TerrainSrg::m_clipmapData.m_clipmapSizeFloat;
+    float clipmapMarginSize = TerrainSrg::m_clipmapData.m_macroClipmapMarginSize;
+
     ClipmapLevel clipmapLevel;
+    float closestLevel = 1.0;
+    // ignore the last level, as it will be used as a fall back default.
+    uint clipmapStackSize = TerrainSrg::m_clipmapData.m_macroClipmapStackSize - 1;
+    for (uint clipmapIndex = 0; clipmapIndex < clipmapStackSize; ++clipmapIndex)
+    {
+        float2 clipmapCenterInWorldSpace = GetMacroClipmapCenterInWorldSpace(clipmapIndex);
+        float clipmapToWorldScale = GetMacroClipmapToWorldScale(clipmapIndex);
+        float extent = (clipmapSize / 2.0 - clipmapMarginSize - ExtendedClipmapMarginSize) * clipmapToWorldScale;
 
-    float2 maxRenderSize = float2(maxRenderDistance, maxRenderDistance);
-    // The top clipmap's render distance
-    float2 minRenderSize = maxRenderSize / pow(clipmapScaleBase, (float)(clipmapStackSize - 1u));
+        float contains = AabbContainsPoint(clipmapCenterInWorldSpace, extent, worldPosition);
+        closestLevel += contains;
+    }
```

Comment:
One thing to point out is we now can specify the scale between levels, not restricted to 2. But we can still use log to locate the level.
My previous implementation used log, but I am afraid the margin might affect the result. This is a set of true world coordinates with margin of 4 texels:
18.18359, 75.54688
18.17188, 75.54688
18.1875, 75.50
18.00, 75.50
17.50, 75.50
16.50, 72.50
Outer clipmaps have bigger difference because their margin is bigger in world coordinates.
Another issue to consider is the center of the camera is not always the same as the highest resolution clipmap's. That means we probably can't use a uniformed center. Though `ExtendedClipmapMarginSize` is designed to help resolving this issue, it seems we need to do some further calculation and test to determine a better `ExtendedClipmapMarginSize`.
I'll mark this and think about a better solution as an optimization later. But I think it should be good for now.

Reply:
ok - fine with this for now since it works.